### DISCUSSION TITLE
Continuous collision checking using FCL (or HPP-FCL)

### DISCUSSION THREAD

**Discussion by rjoomen** [tesseract-robotics/tesseract#1021] [Ideas]
Would it be possible to use FCL for continuous collision checking? FCL does [support this natively](https://github.com/flexible-collision-library/fcl/blob/master/include/fcl/narrowphase/continuous_collision.h), but I'm not how much work it would be to integrate this into Tesseract.

Actual reason I'm asking: I'd like to use HPP-FCL for continuous collision checking, [but the FCL functionality still needs to be ported](https://github.com/humanoid-path-planner/hpp-fcl/issues/421). This is a task I'd be willing to take on, if it is proven to work for Tesseract with FCL.

**Comment by Levi-Armstrong** [comment]
I don't think it would be a lot of work using the existing fcl and the bullet cast checker as a guide.

**Reply by rjoomen** [reply]
Could you give some pointers, as I believe the Bullet cast checker is actually doing the casting, whereas with FCL the library provides it, right?

**Reply by Levi-Armstrong** [reply]
I will take a look at fcl and provide a recommendation on how to approach the implementation.

**Reply by Levi-Armstrong** [reply]
I would start by doing the following. Also you should be able to use the bullet casted collision unit tests for testing the fcl continuous version.

1. copying the fcl_discrete_manager.h/cpp and rename fcl_continuous_manager.h/.cpp
   * Update so it inherits from the continuous base class and you can use the bullet casted implementation for references
   * Will probably need some utility to convert from continuous to regular collision, etc for when objects move from the static broadphase to the dynamic and the other way around, but the bullet casted version should be a good reference on how I manage this when setting active links. 
   *  Need to implement the setCollisionObjectTransform which takes two poses and updates the FCL Continuous Collision Object Motion object.
3. Replace the `dynamic_manage_` broadphase objects with the continuous version found [here](https://github.com/flexible-collision-library/fcl/blob/master/include/fcl/broadphase/broadphase_continuous_collision_manager.h).
4. copy the fcl_collision_object_wrapper.h/cpp and rename fcl_continuous_collision_object_wrapper.h/cpp
   * update this to inherit from the ContinuousCollisionObject and need some minor changes.
